helsingborg-stad/helsingborg-io-sls-api
# Feat/signature status handling

## Feature description
User case (dynamoDn cases table) includes a new attribute - persons. Persons is a collection of users that are related to the case. One or more persons can be required for signing the case before submitting it.

Persons concists of one main person whose role is as the owner of the case - applicant. One secondary person whose role is as a co-applicant in a case. There can also be other roles like children etc.

In this particular situation persons is related to how a Viva case is constructed. The applicant and the co-applicant person object contains a hasSigned attribute and the purpose of this is to make it possible for ie. a smart phone application to implement some sort of signing mechanism where the case is required to be signed by one or more applicants before it's ready to be submitted

## Solution description
Signature is updated in connection with information being updated on a case. This is done by sending an extra attribute 'signature' in the PUT request to the endpoint /cases/{id}.

When a signature is sent the underlying lambda function - updateCase checks the attribute persons that are on a case to validate if the case has all the signatures needed to allow the user to submit the case.

## What areas is affected by these changes?
The cases-api service - updateCase lambda

New attribute 'persons' and 'signature' on cases dynamoDb table

## Is there any exsisting behavior change of other features due to this code change?
The endpoint /cases/{id} request body is in someway altered. Required PUT request body is shown below.
```
{
  "currentFormId": "nnnn",
  "currentPosition": {
    ...
  },
  "signature": {
    "success": true|false
  },
  "answers": [...]
}
```

## How to test
Verify attribute GSI1 is set to USER#(COAPPLICANT PNR) on the appalicant case with PK = USER#(APPLICANT PNR)

Make sure case contains the persons attribute
```
persons: [
  {
    personalNumber: (applicant),
    firstName: some-name,
    lastName: some-name,
    hasSigned: false,
    role: applicant
  },
  {
    personalNumber: (coApplicant),
    firstName: some-name,
    lastName: some-name,
    hasSigned: false,
    role: coApplicant
  },
]
```

Case status type set to notStarted
```
{
  type: 'notStarted:viva',
  name: 'Öppen',
  description: 'Ansökan är öppen. Du kan nu söka ekonomiskt bistånd för perioden.',
}
```
Answers should be encrypted or paste in below example answers via dynamoDb console
```
answers: [
  {
    'encryptedAnswers': 'abc123',
  }
]
```

Update case with PUT request via endpoint /case/{id} with the example request body from above and set signature.success to true. Send request using user appalicant access token. Repeate this request using coappalicant access token
```
...
"signature": {
  "success": true
},
...
```
Check case status


## Was this feature tested in the following environments?
